tests/integration/security/sds_ingress/util/util.go

```diff
@@ -0,0 +1,256 @@
+	MultiMTLSGateway  GatewayType = 3
+)
+
+type IngressCredential struct {
+	PrivateKey string
+	ServerCert string
+	CaCert     string
+}
+
+var IngressCredentialA = IngressCredential{
+	PrivateKey: TLSServerKeyA,
+	ServerCert: TLSServerCertA,
+	CaCert:     CaCertA,
+}
+var IngressCredentialB = IngressCredential{
+	PrivateKey: TLSServerKeyB,
+	ServerCert: TLSServerCertB,
+	CaCert:     CaCertB,
+}
+
+// CreateIngressKubeSecret reads credential names from credNames and key/cert from ingressCred,
+// and creates K8s secrets for ingress gateway.
+// nolint: interfacer
+func CreateIngressKubeSecret(t *testing.T, ctx framework.TestContext, credNames []string,
+	ingressType ingress.IgType, ingressCred IngressCredential) {
+	// Get namespace for ingress gateway pod.
+	istioCfg := istio.DefaultConfigOrFail(t, ctx)
+	systemNS := namespace.ClaimOrFail(t, ctx, istioCfg.SystemNamespace)
+
+	if len(credNames) == 0 {
+		t.Log("no credential names are specified, skip creating ingress secret")
+		return
+	}
+	// Create Kubernetes secret for ingress gateway
+	kubeAccessor := ctx.Environment().(*kube.Environment).Accessor
+	for _, cn := range credNames {
+		secret := createSecret(ingressType, cn, systemNS.Name(), ingressCred)
+		err := kubeAccessor.CreateSecret(systemNS.Name(), secret)
+		if err != nil {
+			t.Errorf("Failed to create secret (error: %s)", err)
+		}
+	}
+	// Check if Kubernetes secret is ready
+	maxRetryNumber := 5
+	checkRetryInterval := time.Second * 1
+	for _, cn := range credNames {
+		t.Logf("Check ingress Kubernetes secret %s:%s...", systemNS.Name(), cn)
+		for i := 0; i < maxRetryNumber; i++ {
+			_, err := kubeAccessor.GetSecret(systemNS.Name()).Get(cn, metav1.GetOptions{})
+			if err != nil {
+				time.Sleep(checkRetryInterval)
+			} else {
+				t.Logf("Secret %s:%s is ready.", systemNS.Name(), cn)
+				break
+			}
+		}
+	}
+}
+
+// createSecret creates a kubernetes secret which stores private key, server certificate for TLS ingress gateway.
+// For mTLS ingress gateway, createSecret adds ca certificate into the secret object.
+func createSecret(ingressType ingress.IgType, cn, ns string, ic IngressCredential) *v1.Secret {
+	if ingressType == ingress.Mtls {
+		return &v1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      cn,
+				Namespace: ns,
+			},
+			Data: map[string][]byte{
+				genericScrtCert:   []byte(ic.ServerCert),
+				genericScrtKey:    []byte(ic.PrivateKey),
+				genericScrtCaCert: []byte(ic.CaCert),
+			},
+		}
+	}
+	return &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cn,
+			Namespace: ns,
+		},
+		Data: map[string][]byte{
+			tlsScrtCert: []byte(ic.ServerCert),
+			tlsScrtKey:  []byte(ic.PrivateKey),
+		},
+	}
+}
+
+type ExpectedResponse struct {
+	ResponseCode int
+	ErrorMessage string
+}
+
+// VisitProductPage makes HTTPS request to ingress gateway to visit product page
+func VisitProductPage(ingress ingress.Instance, host string, timeout time.Duration, exRsp ExpectedResponse, t *testing.T) error {
```

Comment:
Why adding `testing.T`? I see sometimes you return error sometimes we do t.Errorf directly?

Reply:
There are t.Logf() and t.Errorf(), and VisitProductPage keeps retrying until timeout, that is when it returns error. Before timeout it does not return, so t.Errorf() is used.